### ISSUE TITLE
UseCase: Fluent style UseCase

### ISSUE THREAD

**Issue by azu** [almin/almin#193]
This idea from [almin-thinking/02-Fluent-UseCase.ts at master · almin/almin-thinking](https://github.com/almin/almin-thinking/blob/master/02-UseCase/02-Fluent-UseCase.ts "almin-thinking/02-Fluent-UseCase.ts at master · almin/almin-thinking").

```js
new UseCaseExecutor(new MyUseCase())
    .executor(useCase => useCase.execute("test"))
    .then(value => {
        console.log(value);
    });
```

I called it "Fluent style UseCase".
It has some advantages:

### Pros

- Type safe
- Maybe extensible #2 
- Make exist `execute()` catchable(try-catch)

**Type safe**

I've created a very initial POC of fluent syle UseCase.
https://github.com/almin/almin/pull/194

It is actually type safe.

```ts
class MyUseCase extends UseCase {
    execute(value: string) {
        this.dispatch({
            type: "ChildUseCase",
            value
        });
    }
}
context.useCase(new MyUseCase())
    .executor(useCase => useCase.execute(42))
    .then(() => {
        console.log("test");
    });
```

![almin-test ts - almin 2017-05-10 00-12-17](https://cloud.githubusercontent.com/assets/19714/25857930/6e565ac6-3515-11e7-9def-1d045c46b27a.png)

> **Active** middleware that has a side effect is needed to plugin/mixin system.
> It does write too.
> For example, modify `Context#useCase` and make it accepting custom object which is not a UseCase instance.
> 
> We will focus on **Active** midldleware.
> https://github.com/almin/almin/issues/2#issuecomment-290269856

I think that this fluent style may be extensible.

This wrap `UseCase` and pass it to `executor(useCase => )`.

- https://github.com/almin/almin/pull/194/files#diff-2b5be7b20e203152b1265535844e58efR207

We can add middleware system to this wrapping.

### Cons

- A bit complex syntax?
- Conflict with current UseCaseExecutor#execute

If we will add middleware to UseCase, `UseCaseExecutor#execute` and `UseCaseExecutor#executor` should be one method.

----

I think that `executor(executorFunction)` is not good name.
Make it more meaningful name...

Any idea?

**Comment by azu** [comment]
> If we will add middleware to UseCase, UseCaseExecutor#execute and UseCaseExecutor#executor should be one method.

I've changed `UseCaseExecutor#execute` as folloinwg:

```ts
execute(...args: Array<any>): Promise<void> {
        return this.executor((useCase) => useCase.execute(...args));
}
```

It make `execute()` syntax suger of `executor`.

**Comment by azu** [comment]
> Make exist execute() catchable(try-catch)

I've merged #194 for improving error handing on UseCase.
But, it does not means that `executor` is stable.(It maybe changed until next release)

We continue to investigate this Fluent style UseCase.

**Comment by azu** [comment]
Move to https://github.com/almin/almin/issues/107
